### PR TITLE
Vfs reset cache

### PR DESCRIPTION
**Pull request by leoluan1**
#### What is the purpose of this change?

This patch detects out-of-space errors of the full cache mode and resets clean cache files (not changed) to make space.

#### Was the change discussed in an issue or in the forum before?

https://forum.rclone.org/t/removing-recreating-cache-file-upon-enospc/17791

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
- [ ] I have added tests for all changes in this PR if appropriate.
- [ ] I have added documentation for the changes if appropriate.
- [ ] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [ ] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

File: vfs/vfscache/cache.go [hunk 1]
```diff
@@ -526,23 +598,27 @@ func (c *Cache) clean() {
 	fs.Infof(nil, "vfs cache: cleaned: objects %d (was %d) in use %d, to upload %d, uploading %d, total size %v (was %v)", newItems, oldItems, totalInUse, uploadsQueued, uploadsInProgress, newUsed, oldUsed)
 }
 
-// cleaner calls clean at regular intervals
+// cleaner calls clean at regular intervals and upon being kicked for out-of-space condition
 //
 // doesn't return until context is cancelled
 func (c *Cache) cleaner(ctx context.Context) {
+	c.kick = make(chan struct{}, 1)
```
Comment (ncw) [inline vfs/vfscache/cache.go:605]: I would initialise both of these in the main cache initialization `New` so we know they are valid immediately as the `go c.cleaner()` could conceivably run after something uses `c.kick` or `c.cond`

File: vfs/vfscache/downloaders/downloaders.go [hunk 2]
```diff
@@ -386,6 +389,11 @@ func (dls *Downloaders) kickWaiters() (err error) {
 			fs.Errorf(dls.src, "vfs cache: restart download failed: %v", err)
 		}
 	}
+	if fserrors.IsErrNoSpace(dls.lastErr) {
+		fs.Errorf(dls.src, "vfs cache: cache is out of space %d/%d: last error: %v", dls.errorCount, maxErrorCount, dls.lastErr)
+		dls._closeWaiters(dls.lastErr)
```
Comment (ncw) [inline vfs/vfscache/downloaders/downloaders.go:394]: This returns the error to the waiters - I thought you wanted to avoid that?

File: vfs/vfscache/downloaders/downloaders.go [hunk 3]
```diff
@@ -576,9 +584,15 @@ func (dl *downloader) stopAndClose(inErr error) (err error) {
 func (dl *downloader) download() (n int64, err error) {
 	// defer log.Trace(dl.dls.src, "")("err=%v", &err)
 	n, err = dl.in.WriteTo(dl)
-	if err != nil && errors.Cause(err) != asyncreader.ErrorStreamAbandoned {
-		return n, errors.Wrap(err, "vfs reader: failed to write to cache file")
+	if err != nil {
+		if fserrors.IsErrNoSpace(err) {
```
Comment (ncw) [inline vfs/vfscache/downloaders/downloaders.go:588]: I'm not sure you need this bit of code at all - `IsErrNoSpace()` should be able to unwrap the error just fine so I don't see the purpose?

File: vfs/vfscache/item.go [hunk 4]
```diff
@@ -399,9 +402,37 @@ func (item *Item) IsDirty() bool {
...
+func (item *Item) IsDataDirty() bool {
+	item.mu.Lock()
+	defer item.mu.Unlock()
+	return item.info.Dirty
+}
+
+// Open the local file from the object passed in.  Wraps open()
+// to provide recovery from out of space error.
+func (item *Item) Open(o fs.Object) (err error) {
+	for {
```
Comment (ncw) [inline vfs/vfscache/item.go:415]: This probably shouldn't retry forever...

File: vfs/vfscache/item.go [hunk 5]
```diff
@@ -429,6 +460,7 @@ func (item *Item) Open(o fs.Object) (err error) {
 
 	fd, err := file.OpenFile(osPath, os.O_RDWR, 0600)
 	if err != nil {
+		item.opens--
```
Comment (ncw) [inline vfs/vfscache/item.go:463]: None of the other error handling does `item.opens--` so this sticks out. Maybe they all should, I'm not sure...

File: vfs/vfscache/item.go [hunk 6]
```diff
@@ -439,7 +471,14 @@ func (item *Item) Open(o fs.Object) (err error) {
 
 	err = item._save()
 	if err != nil {
-		return err
+		closeErr := item.fd.Close()
+		if closeErr != nil {
+			closeErr = fs.CountError(closeErr)
+			log.Fatal(closeErr)
```
Comment (ncw) [inline vfs/vfscache/item.go:477]: You don't want `log.Fatal` this kills the program! Just `fs.Errorf` the error

File: vfs/vfscache/item.go [hunk 7]
```diff
@@ -767,6 +806,165 @@ func (item *Item) remove(reason string) (wasWriting bool) {
...
+			fs.Errorf(item.o, "vfs cache: item reset failed: %v", e)
+			if err == nil {
+				err = e
+			}
+		}
+	}
+
+	if downloaders := item.downloaders; downloaders != nil {
+		item.downloaders = nil
+		// FIXME need to unlock to kill downloader - should we
```
Comment (ncw) [inline vfs/vfscache/item.go:866]: Mmm, possibly.  The locking is super tricky here. It is really easy to cause a deadlock so if we want to do that then it should be a different commit.

File: vfs/vfscache/item.go [hunk 8]
```diff
@@ -767,6 +806,165 @@ func (item *Item) remove(reason string) (wasWriting bool) {
...
+
+	// item.fd should be nil at this time
+	if item.fd != nil {
+		checkErr(errors.New("vfs cache item: internal error: didn't Close file"))
+		item.fd = nil
+		return
+	}
+	item.modified = false
+
+	fd, err := file.OpenFile(osPath, os.O_RDWR, 0600)
```
Comment (ncw) [inline vfs/vfscache/item.go:915]: This code is duplicated from `Open` - we should probably factor it out of there rather than repeat it...

File: vfs/read_write.go [hunk 9]
```diff
@@ -260,7 +261,21 @@ func (fh *RWFileHandle) _readAt(b []byte, off int64, release bool) (n int, err e
 		// Do the writing with fh.mu unlocked
 		fh.mu.Unlock()
 	}
-	n, err = fh.item.ReadAt(b, off)
+	for {
+		fh.item.ProtectCache()
```
Comment (ncw) [inline vfs/read_write.go:265]: I'm not a big fan of the `ProtectCache` calls - they are an implementation detail for the cache, so can we push them into the item.ReadAt call and any others which need it?

File: vfs/read_write.go [hunk 10]
```diff
@@ -260,7 +261,21 @@ func (fh *RWFileHandle) _readAt(b []byte, off int64, release bool) (n int, err e
...
+		n, err = fh.item.ReadAt(b, off)
+		fh.item.UnprotectCache()
+		if err == nil {
+			break
+		}
+		log.Trace(fh.logPrefix(), "err=%v err.Error()=%v", &err, err.Error())
+		if !fserrors.IsErrNoSpace(err) && err.Error() != "no space left on device" {
```
Comment (ncw) [inline vfs/read_write.go:272]: Likewise this logic should be in Item I think, then you could get rid of `item.GetCache`

File: vfs/vfscache/item.go [hunk 11]
```diff
@@ -811,6 +1009,10 @@ func (item *Item) _ensure(offset, size int64) (err error) {
 	}
 	r := ranges.Range{Pos: offset, Size: size}
 	present := item.info.Rs.Present(r)
+	/* This statement simulates a cache space error for test purpose */
```
Comment (ncw) [inline vfs/vfscache/item.go:1012]: This could maybe be a private cache option that we can set during tests

File: vfs/vfscache/item_test.go [hunk 12]
```diff
@@ -173,7 +173,7 @@ func TestItemTruncateExisting(t *testing.T) {
 	require.Error(t, item.Truncate(40))
 	checkObject(t, r, "existing", contents)
 
-	require.NoError(t, item.Open(obj))
```
Comment (ncw) [inline vfs/vfscache/item_test.go:176]: I'm not sure the purpose of this change - I would have thought the test should work without it.

File: vfs/vfscache/cache.go [hunk 13]
```diff
@@ -511,6 +571,18 @@ func (c *Cache) clean() {
 	// oldest first
 	c.purgeOverQuota(int64(c.opt.CacheMaxSize))
 
+	// removeCleanFiles indicates that we got ENOSPC error
```
[the remark this answers is not in the document]
Comment (leoluan1) [reply vfs/vfscache/cache.go:574]: In our use case, the cache size will be relatively small and it will be filled quickly, like in a few seconds.  So we probably want delete all clean files so that the cleaner does not get kicked even more quickly to repeat the cache reset process.  I think we can achieve this by setting a very low quota in our use case and allow other use cases to reset just enough clean cache files to make the total cache space usage to be below the quota.  I will add the quota check in purgeClean().

File: vfs/vfscache/cache.go [hunk 14]
```diff
@@ -401,6 +407,60 @@ func (c *Cache) reload(ctx context.Context) error {
...
+	c.mu.Lock()
+	for c.outOfSpace == true {
+		fs.Debugf(nil, "vfs cache: in KickCleaner, looping on c.outOfSpace")
+		c.cond.Wait()
+	}
+	fs.Debugf(nil, "vfs cache: in KickCleaner, leaving c.outOfSpace loop")
+	c.mu.Unlock()
+}
+
+func (c *Cache) purgeClean() {
```
[the remark this answers is not in the document]
Comment (leoluan1) [reply vfs/vfscache/cache.go:437]: Can we add this in a subsequent pull request?  I hope to be able to catch the current release.  So minimizing the changes to an essential patch for this pull request might help?

File: vfs/vfscache/cache.go (hunk 1 above)
Comment (leoluan1) [reply vfs/vfscache/cache.go:605]: Makes sense!  Will do.

File: vfs/vfscache/downloaders/downloaders.go (hunk 2 above)
Comment (leoluan1) [reply vfs/vfscache/downloaders/downloaders.go:394]: Don't we need the ENOSPC error to be passed back to the waiters so that the item._ensure() can get the error to start cache reset?

File: vfs/read_write.go (hunk 9 above)
Comment (leoluan1) [reply vfs/read_write.go:265]: Agreed.  I also noticed that and will move them into item.ReadAt.

File: vfs/read_write.go (hunk 10 above)
Comment (leoluan1) [reply vfs/read_write.go:272]: Agreed!

File: vfs/vfscache/item.go (hunk 4 above)
Comment (leoluan1) [reply vfs/vfscache/item.go:415]: Use a maximum retry count of 20 for this commit and potentially add a timeout in the future?

File: vfs/vfscache/downloaders/downloaders.go (hunk 3 above)
Comment (leoluan1) [reply vfs/vfscache/downloaders/downloaders.go:588]: The original code does not return non-nil error (or the ENOSPC error that we are considering currently) unless errors.Cause(err) != asyncreader.ErrorStreamAbandoned.
```
if err != nil && errors.Cause(err) != asyncreader.ErrorStreamAbandoned {
    return n, errors.Wrap(err, "vfs reader: failed to write to cache file")
}
return n, nil
```
Maybe that's bug?  Or I am missing something.
Should the line be 
```
return n, err
```

File: vfs/vfscache/item.go (hunk 8 above)
Comment (leoluan1) [reply vfs/vfscache/item.go:915]: How about factoring the following line out from item.open() and into a new item._newCacheFile() function and call it from both item.open() and item.Reset()?

```
	item.modified = false

	fd, err := file.OpenFile(osPath, os.O_RDWR, 0600)
	if err != nil {
		item.opens--
		return errors.Wrap(err, "vfs cache item: open failed")
	}
	err = file.SetSparse(fd)
	if err != nil {
		fs.Debugf(item.name, "vfs cache: failed to set as a sparse file: %v", err)
	}
	item.fd = fd

	err = item._save()
	if err != nil {
		closeErr := item.fd.Close()
		if closeErr != nil {
			closeErr = fs.CountError(closeErr)
			log.Fatal(closeErr)
		}
		item.fd = nil
		item.opens--
		return errors.Wrap(err, "vfs cache item: _save failed")
	}
```

File: vfs/vfscache/item.go (hunk 11 above)
Comment (leoluan1) [reply vfs/vfscache/item.go:1012]: Sound a great idea! Could you help adding this cache option and the test triggers while I get the function implementation done to get it tested for the current beta?  I would like to do it myself, but it would probably take longer and affect the timeliness of the feature.

File: vfs/vfscache/item_test.go (hunk 12 above)
Comment (leoluan1) [reply vfs/vfscache/item_test.go:176]: As discussed in our PMs, this is due to some incorrect revert of some lines of an earlier commit.  Will fix.

File: vfs/vfscache/item.go (hunk 5 above)
Comment (leoluan1) [reply vfs/vfscache/item.go:463]: I will check all errors under item.Open, item.Close, and item.Reset to see if they adjust item.opens correctly and discuss with you upon questions.

File: vfs/vfscache/item.go (hunk 6 above)
Comment (leoluan1) [reply vfs/vfscache/item.go:477]: This was not the intention.  Will fix it accordingly.

File: vfs/vfscache/item.go (hunk 7 above)
Comment (leoluan1) [reply vfs/vfscache/item.go:866]: Are you referring to the comment in Line 866?  This is actually your comment that got copied from item.close since we are reusing a part of the item.close code.  I do not try to address this comment now. :)

File: vfs/vfscache/downloaders/downloaders.go (hunk 2 above)
Comment (ncw) [reply vfs/vfscache/downloaders/downloaders.go:394]: That makes sense - ignore the above then :-)

File: vfs/vfscache/downloaders/downloaders.go (hunk 3 above)
Comment (ncw) [reply vfs/vfscache/downloaders/downloaders.go:588]: The original code says in English, return an error unless it is ErrorStreamAbandoned in which case don't.

Your new code does exactly the same thing. However I think the original code would work just fine if `err == ENOSPC` then it would be returned in this line

    return n, errors.Wrap(err, "vfs reader: failed to write to cache file")

So I think the major difference between your code and the original is that the ENOSPC error doesn't get wrapped in the `errors.Wrap()` call. I don't think that matters since we are unwrapping the errors to find the ENOSPC anyway so one more layer of wrap doesn't make any difference.

So I think your change works fine, it just isn't needed.

Comment (leoluan1) [reply vfs/vfscache/downloaders/downloaders.go:588]: You are right!  I miss-read the second part of the if condition as "==" (instead of !=).
if err != nil && errors.Cause(err) **==** asyncreader.ErrorStreamAbandoned {

File: vfs/vfscache/cache.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (ncw) [reply vfs/vfscache/cache.go:437]: A subsequent PR is fine

File: vfs/vfscache/item.go (hunk 4 above)
Comment (ncw) [reply vfs/vfscache/item.go:415]: Probably `fs.Config.LowLevelRetries` is the right number (which defaults to 10)

File: vfs/vfscache/item.go (hunk 8 above)
Comment (ncw) [reply vfs/vfscache/item.go:915]: Sounds good :-)

File: vfs/vfscache/item.go (hunk 11 above)
Comment (ncw) [reply vfs/vfscache/item.go:1012]: OK carry on with the PR and at some point I'll merge into a branch we can both work on.

There is bound to be some stuff I'd like to add and that is quite a good way of working.

**Comment by ncw** [comment]
Let me know when you are ready for me to take another look.

**Comment by leoluan1** [comment]
> Let me know when you are ready for me to take another look.

I am mostly done with this round.  There is one area where I am still searching for an approach.   For the cache files that are still dirty after being closed (i.e., they are during the asynchronous write-back, I guess), what can we do about them?  Maybe we should skip the ones with dirty data (item.info.Dirty) and force removing the (item.metaDirty) ones with dirty cache info (item.info)?

File: vfs/vfscache/item.go (hunk 11 above)
Comment (leoluan1) [reply vfs/vfscache/item.go:1003]: > OK carry on with the PR and at some point I'll merge into a branch we can both work on.
> 
> There is bound to be some stuff I'd like to add and that is quite a good way of working.

I think this is a good idea!

**Comment by ncw** [comment]
> I am mostly done with this round. There is one area where I am still searching for an approach. For the cache files that are still dirty after being closed (i.e., they are during the asynchronous write-back, I guess), what can we do about them in the item.Reset(). Maybe we should skip the ones with dirty data (item.info.Dirty) and force removing the (item.metaDirty) ones with dirty cache info (item.info)?

Cache files that are still dirty we shouldn't be touching as if we delete them we'll cause data loss.

Dirty metadata isn't so much of a problem as it is usually quite small. It does take up disk space though and if the disk is full then we can't save it. If the metadata is dirty and we want to remove the file then that won't cause data loss so is no big deal.

I think you'll need to rebase this on master to get the build passing again as go1.15rc2 appears to have vanished...

**Comment by leoluan1** [comment]
> > I am mostly done with this round. There is one area where I am still searching for an approach. For the cache files that are still dirty after being closed (i.e., they are during the asynchronous write-back, I guess), what can we do about them in the item.Reset(). Maybe we should skip the ones with dirty data (item.info.Dirty) and force removing the (item.metaDirty) ones with dirty cache info (item.info)?
> 
> Cache files that are still dirty we shouldn't be touching as if we delete them we'll cause data loss.

I am not touching the cache files with dirty data.

> Dirty metadata isn't so much of a problem as it is usually quite small. It does take up disk space though and if the disk is full then we can't save it. If the metadata is dirty and we want to remove the file then that won't cause data loss so is no big deal.

In my tests I did run into the cases where there is not space for creating a new metadata file after the original cache data and metadata files are removed.  Now I put them on a new errItems field as a part of the cache struct and retry resetting these items after more space if made available.  It's being tested now.  The hope is that we don't need to fail the IOs that already have the file open even after item.Reset fails due to the lack of the small space to save the item's metadata file.

> I think you'll need to rebase this on master to get the build passing again as go1.15rc2 appears to have vanished...

Will do.  Thanks for the heads up.

The refactoring changes in cache.go cause some cache_test.go cases to fail because they need to change as well.   I wonder whether we could work together on those after our branch merge (assuming it's still the plan)?

**Comment by leoluan1** [comment]
> > I am mostly done with this round. There is one area where I am still searching for an approach. For the cache files that are still dirty after being closed (i.e., they are during the asynchronous write-back, I guess), what can we do about them in the item.Reset(). Maybe we should skip the ones with dirty data (item.info.Dirty) and force removing the (item.metaDirty) ones with dirty cache info (item.info)?
> 
> Cache files that are still dirty we shouldn't be touching as if we delete them we'll cause data loss.
> 
> Dirty metadata isn't so much of a problem as it is usually quite small. It does take up disk space though and if the disk is full then we can't save it. If the metadata is dirty and we want to remove the file then that won't cause data loss so is no big deal.
> 
> I think you'll need to rebase this on master to get the build passing again as go1.15rc2 appears to have vanished...

What git steps should I take here?   How does go1.15rc2 impact here?  I am lost here. :)

**Comment by darthShadow** [comment]
> What git steps should I take here?

Update the master branch in your fork and rebase from it to your private branch via `git rebase -p master`. Make sure you have a diff of the changes backed up somewhere in case something goes wrong.

> How does go1.15rc2 impact here?

Currently, since your branch was from an earlier commit in the current master, it's using go1.15rc2 for the build pipeline which is no longer available since 1.15 has been released publicly. Hence the rebase, which will update the pipeline to use 1.15x (the latest 1.15 version available).

**Comment by ncw** [comment]
The build is failing because of these errors

```
vfs/vfscache/cache_test.go:289:17: cannot use removeFile (type func(*Item)) as type func(*Item, time.Duration) in argument to c._purgeOld
vfs/vfscache/cache_test.go:305:17: cannot use removeFile (type func(*Item)) as type func(*Item, time.Duration) in argument to c._purgeOld
vfs/vfscache/cache_test.go:313:17: cannot use removeFile (type func(*Item)) as type func(*Item, time.Duration) in argument to c._purgeOld
vfs/vfscache/cache_test.go:323:16: cannot use removeFile (type func(*Item)) as type func(*Item, time.Duration) in argument to c._purgeOld
vfs/vfscache/cache_test.go:331:17: cannot use removeFile (type func(*Item)) as type func(*Item, time.Duration) in argument to c._purgeOld
```

So the existing tests in `vfscache` need fixing.

**Comment by leoluan1** [comment]
> > What git steps should I take here?
> 
> Update the master branch in your fork and rebase from it to your private branch via `git rebase -p master`. Make sure you have a diff of the changes backed up somewhere in case something goes wrong.
> 
> > How does go1.15rc2 impact here?
> 
> Currently, since your branch was from an earlier commit in the current master, it's using go1.15rc2 for the build pipeline which is no longer available since 1.15 has been released publicly. Hence the rebase, which will update the pipeline to use 1.15x (the latest 1.15 version available).

Thanks for the explanation.  I moved the fork's master to the base's master.  Things look correct and the branch is mergeable again now.

**Comment by leoluan1** [comment]
> The build is failing because of these errors
> 
> ```
> vfs/vfscache/cache_test.go:289:17: cannot use removeFile (type func(*Item)) as type func(*Item, time.Duration) in argument to c._purgeOld
> vfs/vfscache/cache_test.go:305:17: cannot use removeFile (type func(*Item)) as type func(*Item, time.Duration) in argument to c._purgeOld
> vfs/vfscache/cache_test.go:313:17: cannot use removeFile (type func(*Item)) as type func(*Item, time.Duration) in argument to c._purgeOld
> vfs/vfscache/cache_test.go:323:16: cannot use removeFile (type func(*Item)) as type func(*Item, time.Duration) in argument to c._purgeOld
> vfs/vfscache/cache_test.go:331:17: cannot use removeFile (type func(*Item)) as type func(*Item, time.Duration) in argument to c._purgeOld
> ```
> 
> So the existing tests in `vfscache` need fixing.

I fixed the errors related to purge.  However item_test.go have some errors.  Is there any pending bug causing those errors?  I guess not, but just to make sure while I look into them.

**Comment by ncw** [comment]
I think this code is looking very good now and ready to merge :-)

Can you squash the commits - it is probably use one commit isn't it?

Can you also rename the commit(s) so the first line starts with "vfs: " - the change log generator knows what to do with them - [see guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages)

I'd like to get this merged in the next couple of days for an end of the month 1.53 release.

Thank you

**Comment by leoluan1** [comment]
> I think this code is looking very good now and ready to merge :-)
> 
> Can you squash the commits - it is probably use one commit isn't it?
> 
> Can you also rename the commit(s) so the first line starts with "vfs: " - the change log generator knows what to do with them - [see guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages)
> 
> I'd like to get this merged in the next couple of days for an end of the month 1.53 release.
> 
> Thank you

That's great!  Thanks!

BTW, my tests have been running 56 hours continuously over the weekend without encountering deadlock, indefinite hang, or internal errors for reads.   Internal errors for out-of-space writes do come up occasionally by design while all the reads continue happily.

I will squash the commits and rename the commit following the guideline.

**Comment by leoluan1** [comment]
> I think this code is looking very good now and ready to merge :-)
> 
> Can you squash the commits - it is probably use one commit isn't it?
> 
> Can you also rename the commit(s) so the first line starts with "vfs: " - the change log generator knows what to do with them - [see guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages)
> 
> I'd like to get this merged in the next couple of days for an end of the month 1.53 release.
> 
> Thank you

Before squashing the commits, I pushed a small code cleanup fix and ran into a github test error on Mac.  It does not appear to be related to what I changed.  I could not reproduce it on my own Mac either.   Any insight?

The errors are like the following.
```--- FAIL: TestAddPlugin (0.58s)
    rc_test.go:42: 
        	Error Trace:	rc_test.go:42
        	            				rc_test.go:86
        	Error:      	Expected nil, but got: &errors.errorString{s:"could not find an asset in the release. check if asset was successfully added in github release assets"}
        	Test:       	TestAddPlugin
    rc_test.go:88: 
        	Error Trace:	rc_test.go:88
        	Error:      	Should be true
        	Test:       	TestAddPlugin
```

**Comment by ncw** [comment]
I think it all looks ok!

I'll merge this now and hopefully we'll get a bit of testing in before the release.

Thank you very much for your contribution 😃

**Comment by leoluan1** [comment]
That's great!  It's a pleasure to work with you and contribute to this important project!
